From: Lorenzo Stoakes <lstoakes@gmail.com>
Subject: mm: vmalloc: correct use of __GFP_NOWARN mask in __vmalloc_area_node()
Date: Fri, 16 Dec 2022 23:46:59 +0000

This function invokes warn_alloc() with __GFP_NOWARN set which is a no-op.
Set this flag _after_ this call so it is actually invoked, and
additionally remove a duplicate application of __GFP_NOWARN afterwards.

Link: https://lkml.kernel.org/r/20221216234659.493794-1-lstoakes@gmail.com
Signed-off-by: Lorenzo Stoakes <lstoakes@gmail.com>
Cc: Baoquan He <bhe@redhat.com>
Cc: Christoph Hellwig <hch@infradead.org>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Nicholas Piggin <npiggin@gmail.com>
Cc: Uladzislau Rezki <urezki@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/vmalloc.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

--- a/mm/vmalloc.c~mm-vmalloc-correct-use-of-__gfp_nowarn-mask-in-__vmalloc_area_node
+++ a/mm/vmalloc.c
@@ -3031,7 +3031,7 @@ static void *__vmalloc_area_node(struct
 	int ret;
 
 	array_size = (unsigned long)nr_small_pages * sizeof(struct page *);
-	gfp_mask |= __GFP_NOWARN;
+
 	if (!(gfp_mask & (GFP_DMA | GFP_DMA32)))
 		gfp_mask |= __GFP_HIGHMEM;
 
@@ -3051,10 +3051,12 @@ static void *__vmalloc_area_node(struct
 		return NULL;
 	}
 
+	gfp_mask |= __GFP_NOWARN;
+
 	set_vm_area_page_order(area, page_shift - PAGE_SHIFT);
 	page_order = vm_area_page_order(area);
 
-	area->nr_pages = vm_area_alloc_pages(gfp_mask | __GFP_NOWARN,
+	area->nr_pages = vm_area_alloc_pages(gfp_mask,
 		node, page_order, nr_small_pages, area->pages);
 
 	atomic_long_add(area->nr_pages, &nr_vmalloc_pages);
_
